modxcms/revolution
# Refactor template variable creation

### What does it do?
This proposed enhancement changes the way that TV input and output properties are incorporated into the TV editing form, generating these dynamically-loaded fields as Ext components (as the rest of the form does) instead of a rendered DOM element. This allows a level of manipulation of those fields (and native Ext tracking of their dirty states) that's not currently possible, as well as fully-functional validation.

To preserve compatibility with community-developed custom TVs, the original renders and templates were left untouched and a new directory "configs" was created alongside the original "renders" one. The structure within is essentially the same, though the field configs are set up directly in the php files instead of loading a Smarty-parsed tpl file. These configs are subsequently iterated in javascript and inserted as items into a new fieldset component.

Other changes and enhancements include:
- Default Value and Input Option Values fields are only shown when appropriate (this was fixed ahead of this PR, but is implemented in a different way here). 
- Provides a way for user-contributed TVs to hide Default Value and Input Option Values fields (via properties set in the php config, not explicitly illustrated in this PR)
- Default Value field and its labelling changes based on the chosen TV type (e.g., when a file TV is created, its DV field is of the type file and opens the file browser when its trigger is clicked)
- Example code in help descriptions is now subtly highlighted and gets inserted into the relevant field when clicked. Undo functionality is provided by new reset and clear icon buttons (shown next to field label when hovered).
- The actual modx tag for the TV being created now shows in its Name description and gets copied to the clipboard when clicked. The coloring of the clickable text and bg is different than that of the code examples to hint that the expected action is different (copy vs. field insertion).
- Many lexicon entries have been created and/or updated to be more descriptive and precise. Some existing keys were changed to align with the established naming convention ("xxx_desc" for descriptions, "xxx_msg" for messages). I kept references to the originals at the end of each lexicon file edited, just in case these keys are used elsewhere (although I suspect they are not). Note that I wanted to wait to align the non-English language lexicons with this change until it was approved.
- New methods created for form panels to provide a clear path for TV extras (that update to the new loader) to add their own listeners and validators for their field configs to the stock set on the fly (not explicitly illustrated in this PR, but I would create a skeleton example override to show usage).
- Changed date TV input options Start Day and Disabled Days fields to be much more easily usable (should be backward compatible for those who might have saved comma-delimited numbers in the original text fields)
- Created new xcheckboxgroup xtype, used by date Disabled Days, for cb groups that aggregate their checked values into a hidden field (and only save the value of the hidden field to the serialized properties db field). Placed in utilities.js for possible future use elsewhere.
- Minor tweaks to existing SCSS.
- Made many changes to the structure of the field configs to fix layout issues, making all fields stretch to the appropriate widths when the panel's width is changed (also fixes display in mobile context).
- Applied switch/slider style to checkboxes to make consistent with how they appear in Resources. (Note: It probably goes without saying that this would _not_ be done when a checkbox appears in a grid.)
- Groups of fields that only apply in a certain context (e.g., time options for date TV only apply when hideTime is set to "No") only show when appropriate.

### Why is it needed?
To lay the groundwork to solve problems with validation (the dirty state of fields and forms, currently implemented in a very non-ideal way) and refine the control over/display of element form fields (particularly that of TVs). Assuming eventual approval of these changes, applicable ones will then be created along the same guidelines for other elements (chunks, snippets, etc.). The advantages will become even more apparent when my validation fixes for 2.x (from PR #15146) are introduced into 3.x (which I'll turn to once this is _hopefully_ merged).

### How to test
- First create all of the native TV types and confirm expected functionality, both in the TV editing panel and when created TVs are placed within a test Resource. ([Link to a SQL dump of all core TV types](https://github.com/modxcms/revolution/pull/15773#issuecomment-890254926)).
- Install at least one user-contributed TV extra to confirm that the legacy loading functionality works as expected in the TV editing panel.
- Note the highlighted content of the help descriptions, and click to confirm new insertion feature (clicked example gets inserted into field)
- Confirm that fields with insertion feature show the reset and clear utility buttons when hovering over their labels.

### Related issue(s)/PR(s)
Related to #15164, the original draft version of these changes for 2.x.
